File: leetcode/medium/InsertInterval.py
from typing import List

class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # This list will store the final merged intervals.
        result = []
        
        # Iterate through each interval in the input list.
        for i, current_interval in enumerate(intervals):
            # Case 1: The new interval is completely after the current interval.
            # This means the current interval does not overlap and comes before the new one.
            if newInterval[0] > current_interval[1]:
                result.append(current_interval)
            
            # Case 2: The new interval is completely before the current interval.
            # This means we have found the correct insertion point.
            elif newInterval[1] < current_interval[0]:
                # Add the (potentially merged) new interval.
                result.append(newInterval)
                # Then, add all the remaining intervals from the original list,
                # as they are all guaranteed to be after the new interval.
                return result + intervals[i:]
            
            # Case 3: There is an overlap between the new interval and the current one.
            # We merge them by updating the new interval to encompass the current one.
            else:
                newInterval[0] = min(newInterval[0], current_interval[0])
                newInterval[1] = max(newInterval[1], current_interval[1])

        # After the loop, the (potentially merged) new interval still needs to be added.
        # This handles cases where the new interval is inserted at the end of the list
        # or the input `intervals` list was empty.
        result.append(newInterval)
        
        return result

File: leetcode/medium/InsertInterval.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        for (int i = 0; i < intervals.length; i++) {
            int[] currentInterval = intervals[i];
            if (newInterval[0] > currentInterval[1]) {
                result.add(currentInterval);
            }
            else if (newInterval[1] < currentInterval[0]) {
                result.add(newInterval);
                for (int j = i; j < intervals.length; j++) {
                    result.add(intervals[j]);
                }
                return result.stream().toArray(int[][]::new);
            }
            else {
                newInterval[0] = Math.min(newInterval[0], currentInterval[0]);
                newInterval[1] = Math.max(newInterval[1], currentInterval[1]);
            }
        }
        result.add(newInterval);
        return result.stream().toArray(int[][]::new);
    }
}